Source/ASCollectionView.mm

```diff
@@ -622,15 +622,17 @@ - (CGSize)sizeForElement:(ASCollectionElement *)element
     return CGSizeZero;
   }
 
-  NSString *supplementaryKind = element.supplementaryElementKind;
-  NSIndexPath *indexPath = [_dataController.visibleMap indexPathForElement:element];
-  ASSizeRange sizeRange;
-  if (supplementaryKind == nil) {
-    sizeRange = [self dataController:_dataController constrainedSizeForNodeAtIndexPath:indexPath];
+  ASCellNode *node = element.node;
+  BOOL useUIKitCell = node.shouldUseUIKitCell;
+  if (useUIKitCell) {
+    // In this case, we should use the exact value that was stashed earlier by calling sizeForItem:, referenceSizeFor*, etc.
+    // Although the node would use the preferredSize in layoutThatFits, we can skip this because there's no constrainedSize.
+    ASDisplayNodeAssert([node.superclass isSubclassOfClass:[ASCellNode class]] == NO,
+                        @"Placeholder cells for UIKit passthrough should be generic ASCellNodes: %@", node);
+    return node.style.preferredSize;
   } else {
-    sizeRange = [self dataController:_dataController constrainedSizeForSupplementaryNodeOfKind:supplementaryKind atIndexPath:indexPath];
+    return [node layoutThatFits:element.constrainedSize].size;
```

Comment:
Ok, so we no longer reach out to the delegate for a size range but rely on `element.constrainedSize`. This seems like  a simpler approach 👍  

It also means once a size range is returned by the delegate, it's no longer mutable by clients. They have to reload the element to change it, or request for a new API which simply exposes `-[ASDataController relayoutNodes:nodesSizeChanged:]`. Until then, let's get this rolling.